File: rust/src/year2015/day01.rs
use crate::utils::read_chars;

pub fn run() {
    let input = read_chars("../inputs/2015/day01.txt");
    println!("Part 1: {}", part1(&input));
    println!("Part 2: {}", part2(&input));
}

fn part1(input: &[char]) -> i32 {
    input.iter().map(|&c| if c == '(' { 1 } else { -1 }).sum()
}

fn part2(input: &[char]) -> i32 {
    let mut floor = 0;
    for (i, &step) in input.iter().enumerate() {
        if floor < 0 { return i as i32 }
        floor += if step == '(' { 1 } else { -1 };
    }
    panic!("Never entered the basement")
}

File: csharp/Year2015/Day01.cs
namespace csharp.Year2015;

public class Day01 : ISolution {
    public void Solve() {
        List<char> input = [.. File.ReadAllText("../inputs/2015/day01.txt").Trim()];

        Console.WriteLine($"Part 1: {input.Count - 2 * input.Count(item => item == ')')}");
        Console.WriteLine($"Part 2: {GetNegativeStep(input)}");
    }

    private static int GetNegativeStep(List<char> input) {
        int floorNum = 0;
        for (int i = 0; i < input.Count; i++) {
            if (floorNum < 0) {
                return i;
            }

            if (input[i] == '(') {
                floorNum++;
            } else {
                floorNum--;
            }
        }

        return 0;
    }
}
